File: app/src/main/java/com/sorcererxw/matthiasheiderichphotography/ui/adapters/MHAdapter.java
package com.sorcererxw.matthiasheiderichphotography.ui.adapters;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Vibrator;
import android.support.v7.widget.RecyclerView;
import android.util.SparseBooleanArray;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.eftimoff.androipathview.PathView;
import com.mikepenz.materialdrawer.model.BaseViewHolder;
import com.sorcererxw.matthiasheiderichphotography.MHApp;
import com.sorcererxw.matthiasheiderichphotography.ui.activities.DetailActivity;
import com.sorcererxw.matthiasheiderichphotography.ui.others.MHItemAnimInterpolator;
import com.sorcererxw.matthiasheiderichphotography.util.ResourceUtil;
import com.sorcererxw.matthiasheiderichphotography.util.ThemeHelper;
import com.sorcererxw.matthiasheidericphotography.R;
import com.wang.avi.AVLoadingIndicatorView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

import static android.view.View.GONE;

/**
 * @description:
 * @author: Sorcerer
 * @date: 2016/8/22
 */
public class MHAdapter extends RecyclerView.Adapter<MHAdapter.MHBaseViewHolder> {

    private static final boolean FOOTER_ENABLE = false;

    public static class MHBaseViewHolder extends RecyclerView.ViewHolder {

        public MHBaseViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public static class MHFooterViewHolder extends MHBaseViewHolder {

        @BindView(R.id.pathView_foot)
        PathView footView;

        public MHFooterViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class MHViewHolder extends MHBaseViewHolder {
        Drawable drawable;

        @BindView(R.id.pathView_item)
        PathView pathView;

        @BindView(R.id.imageView_item)
        ImageView image;

        @BindView(R.id.loadingIndicator_item)
        AVLoadingIndicatorView loadingIndicatorView;

        public MHViewHolder(View itemView) {
            super(itemView);
        }

        public void playLikeAnim(Context context) {
            ((Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE))
                    .vibrate(150);
            pathView.getPathAnimator()
                    .duration(400)
                    .listenerStart(new PathView.AnimatorBuilder.ListenerStart() {
                        @Override
                        public void onAnimationStart() {
                            pathView.setAlpha(1);
                        }
                    })
                    .listenerEnd(new PathView.AnimatorBuilder.ListenerEnd() {
                        @Override
                        public void onAnimationEnd() {
                            pathView.animate()
                                    .setStartDelay(100)
                                    .alpha(0)
                                    .setDuration(300)
                                    .start();
                        }
                    })
                    .interpolator(new AccelerateDecelerateInterpolator())
                    .start();
        }

        public void playDislikeAnim(Context context) {
            ((Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE))
                    .vibrate(150);
            pathView.animate()
                    .alpha(1)
                    .scaleX(1.2f)
                    .scaleY(1.2f)
                    .setDuration(300)
                    .setListener(new AnimatorListenerAdapter() {
                        @Override
                        public void onAnimationEnd(Animator animation) {
                            pathView.animate()
                                    .alpha(0)
                                    .scaleX(0)
                                    .scaleY(0)
                                    .setDuration(500)
                                    .setListener(new AnimatorListenerAdapter() {
                                        @Override
                                        public void onAnimationEnd(Animator animation) {
                                            super.onAnimationEnd(animation);
                                            pathView.setScaleX(1);
                                            pathView.setScaleY(1);
                                        }
                                    })
                                    .start();
                        }
                    })
                    .start();
        }
    }

    private List<String> mList = new ArrayList<>();
    private Activity mContext;
    private RequestManager mGlideRequestManager;
    private boolean mNightMode;

    public MHAdapter(Activity context) {
        mContext = context;
        mGlideRequestManager = Glide.with(context);
    }

    public void setData(List<String> list) {
        mList = list;
        notifyDataSetChanged();
    }

    @Override
    public MHBaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_FOOTER) {
            return new MHFooterViewHolder(
                    LayoutInflater.from(mContext).inflate(R.layout.item_mh_footer, parent, false));
        }
        return new MHViewHolder(
                LayoutInflater.from(mContext).inflate(R.layout.item_mh, parent, false));
    }

    private SparseBooleanArray mShowedMap = new SparseBooleanArray();

    public boolean hasItemShowed(int position) {
        return mShowedMap.get(position);
    }

    public interface OnItemLongClickListener {
        void onLongClick(View view, String data, int position, MHViewHolder holder);
    }

    private OnItemLongClickListener mOnItemLongClickListener;

    @Override
    public void onBindViewHolder(MHBaseViewHolder baseHolder, int position) {
        if (getItemViewType(position) == TYPE_FOOTER) {
            final MHFooterViewHolder holder = (MHFooterViewHolder) baseHolder;
        } else if (getItemViewType(position) == TYPE_GENERAL) {
            final MHViewHolder holder = (MHViewHolder) baseHolder;
            holder.loadingIndicatorView.setIndicatorColor(ThemeHelper.getAccentColor(mContext));
            holder.itemView.setBackgroundColor(ThemeHelper.getImagePlaceHolderColor(mContext));

            if (mShowedMap.get(position)) {
                holder.loadingIndicatorView.setVisibility(GONE);
            } else {
                holder.loadingIndicatorView.setVisibility(View.VISIBLE);
            }

            Animation animation = new AlphaAnimation(0, 1);
            animation.setDuration(500);
            animation.setInterpolator(new MHItemAnimInterpolator());

            holder.loadingIndicatorView.setAlpha(1);

            mGlideRequestManager.load(mList.get(position))
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .listener(new RequestListener<String, GlideDrawable>() {
                        @Override
                        public boolean onException(Exception e,
                                                   String model,
                                                   Target<GlideDrawable> target,
                                                   boolean isFirstResource) {
                            Timber.e(e);
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(GlideDrawable resource,
                                                       String model,
                                                       Target<GlideDrawable> target,
                                                       boolean isFromMemoryCache,
                                                       boolean isFirstResource) {
                            holder.loadingIndicatorView.animate().alpha(0).start();
                            mShowedMap.put(holder.getAdapterPosition(), true);
                            holder.drawable = resource;
                            return false;
                        }
                    })
                    .animate(animation)
                    .fitCenter()
                    .into(holder.image);

            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Intent intent = new Intent(mContext, DetailActivity.class);
                    intent.putExtra("link", mList.get(holder.getAdapterPosition()));
                    if (mShowedMap.get(holder.getAdapterPosition())) {
                        MHApp.setTmpDrawable(mContext, holder.drawable);
                    } else {
                        MHApp.setTmpDrawable(mContext, null);
                    }
                    mContext.startActivityForResult(intent, 0, ActivityOptions
                            .makeSceneTransitionAnimation(mContext, holder.image, "image")
                            .toBundle());
                }
            });


            holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    if (mOnItemLongClickListener != null) {
                        mOnItemLongClickListener.onLongClick(view,
                                mList.get(holder.getAdapterPosition()),
                                holder.getAdapterPosition(), holder);
                    }
                    return true;
                }
            });
        }
    }

    public void removeItem(int position) {
        mList.remove(position);
        notifyItemRemoved(position);
    }

    public void setNightMode(boolean isNightMode) {
        if (mNightMode == isNightMode) {
            return;
        }
        mNightMode = isNightMode;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        if (mList.size() == 0) {
            return 0;
        }
        if (FOOTER_ENABLE) {
            return mList.size() + 1;
        }
        return mList.size();
    }

    private static final int TYPE_GENERAL = 0x0;
    private static final int TYPE_FOOTER = 0x1;

    @Override
    public int getItemViewType(int position) {
        if (position == getItemCount() - 1 && FOOTER_ENABLE) {
            return TYPE_FOOTER;
        }
        return TYPE_GENERAL;
    }

    public void setOnItemLongClickListener(
            OnItemLongClickListener onItemLongClickListener) {
        mOnItemLongClickListener = onItemLongClickListener;
    }

}

File: app/src/main/java/com/sorcererxw/matthiasheiderichphotography/ui/adapters/MHAdapter.kt
package com.sorcererxw.matthiasheiderichphotography.ui.adapters

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.graphics.Paint
import android.graphics.drawable.Drawable
import android.os.Vibrator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.AlphaAnimation
import android.widget.ImageView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.CircularProgressDrawable
import com.bumptech.glide.load.resource.bitmap.BitmapTransitionOptions
import com.eftimoff.androipathview.PathView
import com.sorcererxw.matthiasheiderichphotography.config.GlideApp
import com.sorcererxw.matthiasheiderichphotography.room.photo.PhotoEntity
import com.sorcererxw.matthiasheiderichphotography.ui.others.MHItemAnimInterpolator
import com.sorcererxw.matthiasheidericphotography.R
import java.util.*

/**
 * @description:
 * @author: Sorcerer
 * @date: 2016/8/22
 */
class MHAdapter(
        private val mContext: Context,
        private var mList: MutableList<PhotoEntity> = ArrayList(),
        var onItemClickListener: OnItemClickListener? = null,
        var onItemLongClickListener: OnItemLongClickListener? = null
) : RecyclerView.Adapter<MHAdapter.MHBaseViewHolder>() {

    private var mNightMode: Boolean = false

    open class MHBaseViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    private class MHFooterViewHolder(itemView: View) : MHBaseViewHolder(itemView) {
        var footView: PathView = itemView.findViewById(R.id.pathView_foot)
    }

    fun setData(newData: List<PhotoEntity>) {
        val diffResult = DiffUtil.calculateDiff(
                MHDiffCallback(mList, newData),
                true
        )
        mList.clear()
        mList.addAll(newData)
        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MHBaseViewHolder {
        return if (viewType == TYPE_FOOTER) {
            MHFooterViewHolder(
                    LayoutInflater.from(mContext).inflate(R.layout.item_mh_footer, parent, false))
        } else MHViewHolder(
                LayoutInflater.from(mContext).inflate(R.layout.item_mh, parent, false))
    }

    interface OnItemClickListener {
        fun onLongClick(view: View, data: PhotoEntity, position: Int, holder: MHViewHolder)
    }

    interface OnItemLongClickListener {
        fun onLongClick(view: View, data: PhotoEntity, position: Int, holder: MHViewHolder)
    }

    override fun onBindViewHolder(baseHolder: MHBaseViewHolder, position: Int) {
        if (getItemViewType(position) == TYPE_FOOTER) {
            val holder = baseHolder as MHFooterViewHolder
        } else if (getItemViewType(position) == TYPE_GENERAL) {
            val holder = baseHolder as MHViewHolder
//            holder.itemView.setBackgroundColor(ThemeHelper.getImagePlaceHolderColor(mContext))
//
//            val animation = AlphaAnimation(0f, 1f)
//            animation.duration = 500
//            animation.interpolator = MHItemAnimInterpolator()

            val indicator = CircularProgressDrawable(mContext);
            indicator.strokeWidth = 10f
            indicator.strokeCap = Paint.Cap.ROUND
            indicator.centerRadius = 50f

            GlideApp.with(mContext)
                    .asBitmap()
                    .load(mList[position].url)
                    .placeholder(indicator)
                    .centerCrop()
                    .transition(BitmapTransitionOptions().crossFade(200))
                    .into(holder.image)

            holder.itemView.setOnClickListener { view ->
                onItemClickListener?.onLongClick(
                        view,
                        mList[holder.adapterPosition],
                        holder.adapterPosition, holder
                )
            }

            holder.itemView.setOnLongClickListener { view ->
                onItemLongClickListener?.onLongClick(
                        view,
                        mList[holder.adapterPosition],
                        holder.adapterPosition, holder
                )
                true
            }
        }
    }

    fun removeItem(position: Int) {
        mList.removeAt(position)
        notifyItemRemoved(position)
    }

    fun setNightMode(isNightMode: Boolean) {
        if (mNightMode == isNightMode) {
            return
        }
        mNightMode = isNightMode
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        if (mList.size == 0) {
            return 0
        }
        return if (FOOTER_ENABLE) {
            mList.size + 1
        } else mList.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == itemCount - 1 && FOOTER_ENABLE) {
            TYPE_FOOTER
        } else TYPE_GENERAL
    }

    private class MHDiffCallback(
            private val mOldList: List<PhotoEntity>,
            private val mNewList: List<PhotoEntity>
    ) : DiffUtil.Callback() {

        override fun getOldListSize() = mOldList.size

        override fun getNewListSize() = mNewList.size

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return mOldList[oldItemPosition].url == mNewList[newItemPosition].url
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return mOldList[oldItemPosition].url == mNewList[newItemPosition].url
        }
    }

    class MHViewHolder(itemView: View) : MHBaseViewHolder(itemView) {
        internal var drawable: Drawable? = null
        val pathView: PathView = itemView.findViewById(R.id.pathView_item)
        val image: ImageView = itemView.findViewById(R.id.imageView_item)

        fun playLikeAnim(context: Context) {
            (context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator).vibrate(150)
            pathView.pathAnimator
                    .duration(400)
                    .listenerStart { pathView.alpha = 1f }
                    .listenerEnd {
                        pathView.animate()
                                .setStartDelay(100)
                                .alpha(0f)
                                .setDuration(300)
                                .start()
                    }
                    .interpolator(AccelerateDecelerateInterpolator())
                    .start()
        }

        fun playDislikeAnim(context: Context) {
            (context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator)
                    .vibrate(150)
            pathView.animate()
                    .alpha(1f)
                    .scaleX(1.2f)
                    .scaleY(1.2f)
                    .setDuration(300)
                    .setListener(object : AnimatorListenerAdapter() {
                        override fun onAnimationEnd(animation: Animator) {
                            pathView.animate()
                                    .alpha(0f)
                                    .scaleX(0f)
                                    .scaleY(0f)
                                    .setDuration(500)
                                    .setListener(object : AnimatorListenerAdapter() {
                                        override fun onAnimationEnd(animation: Animator) {
                                            super.onAnimationEnd(animation)
                                            pathView.scaleX = 1f
                                            pathView.scaleY = 1f
                                        }
                                    })
                                    .start()
                        }
                    })
                    .start()
        }
    }

    companion object {

        private const val FOOTER_ENABLE = false
        private const val TYPE_GENERAL = 0x0
        private const val TYPE_FOOTER = 0x1
    }

}
